src/Processors/Formats/Impl/ParallelFormattingOutputFormat.h

```diff
@@ -0,0 +1,357 @@
+        if (!background_exception)
+        {
+            background_exception = std::current_exception();
+        }
+        emergency_stop = true;
+        writer_condvar.notify_all();
+        collector_condvar.notify_all();
+    }
+
+    void scheduleFormatterThreadForUnitWithNumber(size_t ticket_number)
+    {
+        pool.scheduleOrThrowOnError([this, ticket_number] { formatterThreadFunction(ticket_number); });
+    }
+
+    void collectorThreadFunction()
+    {
+        setThreadName("Collector");
+
+        try
+        {
+            while (!emergency_stop)
+            {
+                const auto current_unit_number = collector_unit_number % processing_units.size();
+                auto & unit = processing_units[current_unit_number];
+
+                {
+                    std::unique_lock<std::mutex> lock(mutex);
+                    collector_condvar.wait(lock,
+                        [&]{ return unit.status == READY_TO_READ || emergency_stop; });
+                }
+
+                if (emergency_stop)
+                    break;
+
+                assert(unit.status == READY_TO_READ);
+
+                /// Use this copy to after notification to stop the execution.
+                auto copy_if_unit_type = unit.type;
+
+                /// Do main work here.
+                out.write(unit.segment.data(), unit.actual_memory_size);
+
+                if (need_flush.exchange(false) || auto_flush)
+                    IOutputFormat::flush();
+
+                ++collector_unit_number;
+
+                {
+                    /// Notify other threads.
+                    std::lock_guard<std::mutex> lock(mutex);
+                    unit.status = READY_TO_INSERT;
+                    writer_condvar.notify_all();
+                }
+                /// We can exit only after writing last piece of to out buffer.
+                if (copy_if_unit_type == ProcessingUnitType::FINALIZE)
+                {
+                    break;
+                }
+            }
+            collector_finished.set();
+        }
+        catch (...)
+        {
+            collector_finished.set();
+            onBackgroundException();
+        }
+    }
+
+
+    void formatterThreadFunction(size_t current_unit_number)
+    {
+        setThreadName("Formatter");
+
+        try
+        {
+            auto & unit = processing_units[current_unit_number];
+            assert(unit.status = READY_TO_FORMAT);
+
+            /// We want to preallocate memory buffer (increase capacity)
+            /// and put the pointer at the beginning of the buffer
+            /// FIXME: Implement reserve() method in Memory.
+            unit.segment.resize(DBMS_DEFAULT_BUFFER_SIZE);
+            unit.segment.resize(0);
+
+            unit.actual_memory_size = 0;
+            BufferWithOutsideMemory<WriteBuffer> out_buffer(unit.segment);
+
+            auto formatter = internal_formatter_creator(out_buffer);
+
+            switch (unit.type)
```

Comment:
It's not clear why reordering of processing types is not possible in current implementation, could you please add a comment?

Reply:
Added a comment. In short, there are 'two pointers' (Collector thread and thread which consumes Chunks) and first is guaranteed to be less or equal than other.